### PR TITLE
Export ClipImageTransform

### PR DESCRIPTION
**Pull request by lucylq**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
- Add torchtune/models/clip/inference/_transforms.py, a version of CLIPImageTransform that can be exported. Update transforms tests with a comparison between the two CLIPImageTransforms. 
- Refactor target_size calculations into function `get_resize_dimensions`, add tests.
- Change tile_crop from a function to an nn.Module for module swapping, update callsites.

Note: we cannot add a test for the exported model, as source transforms are required on the nn.Modules to successfully export it. 

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help. We also have a [contributing page](../CONTRIBUTING.md) for some guidance on contributing.)

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](../tests/torchtune) for any new functionality
- [x] update [docstrings](../docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

### CODE REVIEW

**Review by felipemello1 — COMMENTED** [review]
LGTM, I will leave the final word to @pbontrager. Maybe it would be nice to add a small docstring at the top of "torchtune/models/clip/inference/_transforms.py" explaining that this is for executorch only. But I am fine with leaving it as it is.

File: tests/torchtune/models/clip/test_clip_image_transform.py [hunk 1]
```diff
@@ -11,7 +11,10 @@
 import torch
 from tests.test_utils import assert_expected
 
-from torchtune.models.clip._transforms import CLIPImageTransform
+from torchtune.models.clip._transforms import CLIPImageTransform as CLIPImageTransform
```
[the remark this answers is not in the document]
Comment (lucylq) [reply tests/torchtune/models/clip/test_clip_image_transform.py]: Ah, good catch. Updated

[the remark this answers is not in the document]
Comment (lucylq) [reply tests/torchtune/models/clip/test_clip_image_transform.py]: Also updated - @tarun292 found we can swap out functionals as long as we 'name' them inside __init__(), so leaving torch.nn.functional.pad, and tile_crop as is.

**Comment by lucylq** [comment]
> LGTM, I will leave the final word to @pbontrager. Maybe it would be nice to add a small docstring at the top of "torchtune/models/clip/inference/_transforms.py" explaining that this is for executorch only. But I am fine with leaving it as it is.

Thanks for reviewing! Added a note that being able to export is useful for ExecuTorch/AOTI.

File: torchtune/modules/transforms/vision_utils/get_resize_dimensions.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+from typing import Optional, Tuple
+
+logger = logging.getLogger(__name__)
+
+
+def get_resize_dimensions(
+    image_size: Tuple[int], target_size: Tuple[int], max_upscaling_size: Optional[int]
+) -> Tuple[int]:
+    """
+    Calculates the resize dimensions for an image, given the target size and max_upscaling_size.
```
Comment (pbontrager) [inline torchtune/modules/transforms/vision_utils/get_resize_dimensions.py:20]: I think this could be fleshed out a bit more. The key thing this function does is computes what size an image would be if it was resized to be inscribed within a target size. So it's upscaled or downscaled such that one size equals the target size and the second side is less than or equal to the target size. I think some description describing this would help make the function easier to discover and use.

File: torchtune/modules/transforms/vision_utils/get_resize_dimensions.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+import logging
+
+import math
+
+from typing import Optional, Tuple
+
+logger = logging.getLogger(__name__)
+
+
+def get_resize_dimensions(
```
Comment (pbontrager) [inline torchtune/modules/transforms/vision_utils/get_resize_dimensions.py:16]: This needs to be added to the docs RST

File: torchtune/modules/transforms/vision_utils/get_resize_dimensions.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    Examples:
+
+        Example 1: The image will be upscaled from (300, 800) to (448, 1194), since 448 is the limiting side.
+
+            >>> max_upscaling_size = None
+            >>> image_size = (300, 800)
+            >>> target_size = (448, 1344)
+            >>> output = get_resize_dimensions(image_size, target_size, resample, max_upscaling_size)
+
+        Example 2: The image will stay as is, since 800 > 600.
```
Comment (pbontrager) [inline torchtune/modules/transforms/vision_utils/get_resize_dimensions.py:41]: @felipemello1 I find it counterintuitive that this doesn't cause a downsize to (225, 600). max_upscaling_size is a bit weird to have here at all because you could just not provide a target that's too big, but I understand it's convenient.

**Review by pbontrager — COMMENTED** [review]
This is looking nice and very clean. I really like the docstring for the new transform as well. I left some comments around naming and descriptions but mostly nits. I think the get_resize_dimensions should be added to the docs, but I'm fine with leaving the inference transform out of the docs for now as we figure out our story there.

**Review by pbontrager — COMMENTED** [review]
This is looking nice and very clean. I really like the docstring for the new transform as well. I left some comments around naming and descriptions but mostly nits. I think the get_resize_dimensions should be added to the docs, but I'm fine with leaving the inference transform out of the docs for now as we figure out our story there.

File: torchtune/modules/transforms/vision_utils/get_resize_dimensions.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    image_size: Tuple[int], target_size: Tuple[int], max_upscaling_size: Optional[int]
+) -> Tuple[int]:
+    """
+    Calculates the resize dimensions for an image, given the target size and max_upscaling_size.
+    The user can set max_upscaling_size to limit upscaling when target_size exceeds image_size.
+
+    Args:
+        image_size (Tuple[int]): The size of the image, in the form [height, width].
+        target_size (Tuple[int]): The desired resolution to fit the image into, in the format [height, width].
+        max_upscaling_size (Optional[int]): The maximum size to upscale the image to.
```
[the remark this answers is not in the document]
Comment (lucylq) [reply torchtune/modules/transforms/vision_utils/get_resize_dimensions.py:26]: Sounds good - I'll change this in the other transforms as well so it's consistent.

File: torchtune/modules/transforms/vision_utils/get_resize_dimensions.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lucylq) [reply torchtune/modules/transforms/vision_utils/get_resize_dimensions.py]: sounds good - updated

File: torchtune/modules/transforms/vision_utils/get_resize_dimensions.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        Tuple[int]: The resize dimensions for the image, in the format [height, width].
+
+    Examples:
+
+        Example 1: The image will be upscaled from (300, 800) to (448, 1194), since 448 is the limiting side.
+
+            >>> max_upscaling_size = None
+            >>> image_size = (300, 800)
+            >>> target_size = (448, 1344)
+            >>> output = get_resize_dimensions(image_size, target_size, resample, max_upscaling_size)
```
[the remark this answers is not in the document]
Comment (lucylq) [reply torchtune/modules/transforms/vision_utils/get_resize_dimensions.py]: good catch - updated

File: torchtune/modules/transforms/vision_utils/get_resize_dimensions.py (hunk 3 above)
Comment (lucylq) [reply torchtune/modules/transforms/vision_utils/get_resize_dimensions.py]: thanks!

File: torchtune/modules/transforms/vision_utils/get_resize_dimensions.py (hunk 2 above)
Comment (lucylq) [reply torchtune/modules/transforms/vision_utils/get_resize_dimensions.py]: thanks, updated

**Review by pbontrager — APPROVED** [review]
Thanks for the hard work! I'll approve this now, but the test_get_resize_dimensions test should be renamed before you land it.
